apps/src/blockly/customBlocks/googleBlockly/proceduresBlocks.js

```diff
@@ -187,15 +188,41 @@ GoogleBlockly.Extensions.register('procedures_block_frame', function () {
   }
 });
 
-// Override the destroy function to not destroy the procedure. We need to do this
-// so that when we clear the modal function editor we don't remove the procedure
-// from the procedure map.
+// Override the destroy function to not destroy the procedure if we're using the
+// modal function editor. We need to do this so that when we clear its workspace
+// we don't remove the procedure from the procedure map. Insertion markers also
+// get this destroy function and they should not cause us to delete a procedure
+// either.
 GoogleBlockly.Extensions.register('modal_procedures_no_destroy', function () {
+  const originalDestroy = this.destroy.bind(this);
   const mixin = {
     destroy: function () {
-      // no-op
-      // this overrides the destroy hook registered
-      // in the procedure_def_get_def_mixin
+      if (!Blockly.useModalFunctionEditor && !this.isInsertionMarker()) {
+        originalDestroy();
+      }
+    },
+  };
+  // We can't register this as a mixin since we're overwriting existing methods
+  Object.assign(this, mixin);
+});
+
+// Override the doProcedureUpdate function to heal the stack. Without this,
+// any child blocks connected to a call block would also get deleted.
+GoogleBlockly.Extensions.register('procedure_call_heal_stack', function () {
```

Comment:
nit: `procedure_call_heal_stack` doesn't imply to me that this overrides `doProcedureUpdate`. Could we name it something like `procedure_caller_do_update`?

Reply:
Good call. Updated!